### PR TITLE
bitwarden-cli 1.0.0 (new formula)

### PR DESCRIPTION
**Pull request by kspearrin**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by DomT4** [comment]
Wondering if this should be called `bitwarden-cli` to make it clear it's not installing the same thing as the `bitwarden` cask.

**Comment by kspearrin** [comment]
I could rename it if you think that makes more sense. Our choco package is called `bitwarden-cli` since we already have `bitwarden` for our desktop app there too.

File: Formula/bitwarden.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+  depends_on "node"
+
+  def install
+    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
+    bin.install_symlink Dir["#{libexec}/bin/*"]
+  end
+
+  test do
+    assert_equal "1.0.0\n", shell_output("#{bin}/bw -v")
```
[the remark this answers is not in the document]
Comment (kspearrin) [reply Formula/bitwarden.rb]: Bitwarden is an online password manager that is powered by a server. You would need to register an account.

**Comment by kspearrin** [comment]
Now that I think about it more, I think it would be good to be called `bitwarden-cli` to be more consistent with other packages that we distribute. Renamed.

File: Formula/bitwarden.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (kspearrin) [reply Formula/bitwarden.rb]: Good call. Updated to assert `version.to_s + "\n"`

[the remark this answers is not in the document]
Comment (DomT4) [reply Formula/bitwarden.rb]: If you use `assert_match` rather than `assert_equal` it'll let you off the newline.

File: Formula/bitwarden-cli.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+  depends_on "node"
+
+  def install
+    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
+    bin.install_symlink Dir["#{libexec}/bin/*"]
+  end
+
+  test do
+    assert_equal version.to_s + "\n", shell_output("#{bin}/bw -v")
```
Comment (ilovezfs) [inline Formula/bitwarden-cli.rb]: The test needs to exercise and verify the functionality of the software. Something that saves and returns a password would be good

**Review by ilovezfs — CHANGES_REQUESTED** [review]
see comment on test block

File: Formula/bitwarden-cli.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+  def install
+    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
+    bin.install_symlink Dir["#{libexec}/bin/*"]
+  end
+
+  test do
+    version = shell_output("#{bin}/bw --version")
+    password = shell_output("#{bin}/bw generate --length 10")
+    testingb64 = pipe_output("#{bin}/bw encode", "Testing")
```
Comment (ilovezfs) [inline Formula/bitwarden-cli.rb]: add a `, 0` to assert the exit code is 0 too

File: Formula/bitwarden-cli.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+
+    assert_match version.to_s, version
```
Comment (ilovezfs) [inline Formula/bitwarden-cli.rb]: we don't need this test since you've added the better test

File: Formula/bitwarden-cli.rb [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+    assert_equal 10, password.strip.length
```
Comment (ilovezfs) [inline Formula/bitwarden-cli.rb]: is it only a single newline? I'd just use `.chomp`

**Comment by ilovezfs** [comment]
so probably
```
    assert_equal 10, shell_output("#{bin}/bw generate --length 10").chomp.length
    assert_match "VGVzdGluZw==", pipe_output("#{bin}/bw encode", "Testing", 0)
```

File: Formula/bitwarden-cli.rb [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+  def install
+    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
+    bin.install_symlink Dir["#{libexec}/bin/*"]
+  end
+
+  test do
+    password = shell_output("#{bin}/bw generate --length 10")
+    testingb64 = pipe_output("#{bin}/bw encode", "Testing")
+
+    assert_equal 10, password.chomp.length, 0
```
Comment (ilovezfs) [inline Formula/bitwarden-cli.rb]: sorry I meant as 3rd arg on pipe_output. a 3rd arg to assert_equal is for a custom failure message

**Comment by kspearrin** [comment]
@ilovezfs Done

**Comment by ilovezfs** [comment]
```
    assert_equal 10, shell_output("#{bin}/bw generate --length 10").chomp.length

    output = pipe_output("#{bin}/bw encode", "Testing", 0)
    assert_equal "VGVzdGluZw==", output.chomp
```

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
Thanks for your first contribution to Homebrew, @kspearrin! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**